### PR TITLE
First, basic pass at scanning databinding expressions.

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md not updated, this scanner isn't hooked up anywhere, and still needs more work.

Needed for the linter. For example: https://github.com/Polymer/polymer-linter/issues/29 https://github.com/Polymer/polymer-linter/issues/25 https://github.com/Polymer/polymer-linter/issues/9 and https://github.com/Polymer/polymer-linter/issues/8 will all depend on this work or something much like it.

### CODE REVIEW

**Comment by rictic** [comment]
Sending this out as is while I build on top in the interest of sending out smaller, more easily reviewed chunks.

**Comment by justinfagnani** [comment]
High level comment:

Polymer's expression parser has a lot of quirks. How will we model these against a parser that's much more correct? Will we scan string literals for commas and warn? Will we ban nested expressions?

At some point, when is it easier to model the idiosyncrasies by just using Polymer's set of regexes?

**Comment by TimvdLippe** [comment]
This is probably unrelated to this PR, but instead of adhering to the regex of the core library I think it would be a better idea to refactor these regexes into a small automata in the core library. The automata will be a lot easier to reason about I suppose and does not have the regex quirks.

**Comment by rictic** [comment]
We may want to move to Polymer's regexes at some point. It'd certainly be nice if Polymer core's algorithm was packaged up nicely.

I've thrashed about with different implementations of this a couple times though, and this seemed like a reasonable first pass that moves in the right direction. It won't be identical with to Polymer core's parsing, but we can do fairly well by parsing with espree and then warning based on strict validation of AST shapes.

This gives us a standard and well documented format for the expression parse trees that we can extend out as necessary if more syntax is added at some later date (or if an element uses e.g. polymer-expressions). I think it does not limit us if we later switch to using Polymer core's regexes, as the syntax they match should be a subset of JS.

So in short, I propose this route because it's incremental, adds no dependencies, and doesn't limit us in the medium term.

**Comment by justinfagnani** [comment]
@TimvdLippe 🦄

File: src/javascript/javascript-parser.ts [hunk 1]
```diff
@@ -27,19 +27,21 @@ declare class SyntaxError {
...
     let ast: Program;
-    const options = {
-      ecmaVersion: 8,
-      attachComment: true,
-      comment: true,
-      loc: true,
-      sourceType: 'script' as ('script' | 'module'),
-    };
+    const options = Object.assign(
+        {sourceType: 'script' as ('script' | 'module')}, baseParseOptions);
```
Comment (justinfagnani) [inline src/javascript/javascript-parser.ts]: is that cast necessary? try adding a type to `options`

File: src/polymer/expression-scanner.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: 2017

File: src/polymer/expression-scanner.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+import {HtmlVisitor, ParsedHtmlDocument} from '../html/html-document';
+import {HtmlScanner} from '../html/html-scanner';
+import {baseParseOptions} from '../javascript/javascript-parser';
+import {ScannedFeature, SourceRange} from '../model/model';
+import {Warning} from '../warning/warning';
+
+const p = dom5.predicates;
+const isTemplate = p.hasTagName('template');
+const matchInnerDataBindingTemplate = p.AND(
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: Add comment. I thought at first this was going to include templates in `dom-module`s.

Also, any of these can be used standalone, not sure you need the separate dom-bind predicate.

File: src/polymer/expression-scanner.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        p.parentMatches(p.OR(
+            p.hasTagName('dom-bind'),
+            p.hasTagName('dom-if'),
+            p.hasTagName('dom-repeat')))));
+export const isDomBindTemplate =
+    p.AND(p.hasTagName('template'), p.hasAttrValue('is', 'dom-bind'));
+export const isDomModuleTemplate = p.AND(
+    p.parentMatches(p.hasTagName('dom-module')), p.hasTagName('template'));
+
+export interface Template extends parse5.ASTNode { content: parse5.ASTNode; }
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: If you add tagName: 'template', will that aid narrowing?

File: src/polymer/expression-scanner.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+ * Databinding into an attribute/property.
+ */
+export class ScannedAttributeExpression implements ScannedFeature {
+  /** The element whose attribute is being assigned to.*/
+  astNode: parse5.ASTNode;
+  /** The HTML element attribute that's being assigned to.*/
+  attribute: parse5.ASTAttribute;
+  sourceRange: SourceRange;
+  warnings: Warning[] = [];
+  directionality: 'down-only'|'bidirectional';
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: these are new terms to describe this. I'd go with something like bindingType: one-way/two-way or '{' and '[' (which is how this is represented in core).

File: src/polymer/expression-scanner.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  expressionText: string;
+  expressionAst: estree.Program;
+
+  /**
+   * If this is a bidirectional data binding, and an event name was specified
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: Seems like you should capture the event name even for one-way bindings, since it might be written that way. Lint can warn on it.

File: src/polymer/expression-scanner.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    const results: ScannedAttributeExpression[] = [];
+    const dataBindingTemplates = getAllDataBindingTemplates(document.ast);
+    for (const template of dataBindingTemplates) {
+      dom5.nodeWalkAll(template.content, (node) => {
+        if (node.attrs) {
+          for (const attr of node.attrs) {
+            if (!attr.value) {
+              continue;
+            }
+            const oneWayMatch = attr.value.match(/^\s*\[\[\s*(.*?)\s*\]\]\s*$/);
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: How similar is this to the core regex? how about one regex and check the value of the group?

File: src/polymer/expression-scanner.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+              results.push(new ScannedAttributeExpression(
+                  node,
+                  attr,
+                  document.sourceRangeForAttributeValue(node, attr.name)!,
+                  'down-only',
+                  expressionText,
+                  undefined));
+            } else if (bidiMatch) {
+              let expressionText = bidiMatch[1];
+              const match = expressionText.match(/(.*)::(.*)/);
```
Comment (justinfagnani) [inline src/polymer/expression-scanner.ts]: see comment above, but I'd always try to parse the event

File: src/test/polymer/expression-scanner_test.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
```
Comment (justinfagnani) [inline src/test/polymer/expression-scanner_test.ts]: 2017

File: src/javascript/javascript-parser.ts (hunk 1 above)
Comment (rictic) [inline src/javascript/javascript-parser.ts]: It is, otherwise sourceType is inferred as `string`. Fixed in a later commit.

File: src/polymer/expression-scanner.ts (hunk 2 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Done

File: src/polymer/expression-scanner.ts (hunk 3 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Oh, you can use dom-if and dom-repeat as top level? Well that simplifies things a lot!

File: src/polymer/expression-scanner.ts (hunk 4 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Sadly, no, as parse5.ASTNode's tagName is `string` so it doesn't know that 'template' would imply Template, only that all Templates have 'template'.

File: src/polymer/expression-scanner.ts (hunk 5 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Went with `{` and `[`

File: src/polymer/expression-scanner.ts (hunk 6 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Doing it this way, you get a javascript parse error if you use `::` on anything but a bidirectional match of an entire attribute. No need for the additional lint pass.

File: src/polymer/expression-scanner.ts (hunk 7 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: In the next PR this is significantly refactored and made more efficient to match inline expressions.

File: src/polymer/expression-scanner.ts (hunk 8 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: See comment above.

File: src/test/polymer/expression-scanner_test.ts (hunk 9 above)
Comment (rictic) [inline src/test/polymer/expression-scanner_test.ts]: done

File: src/polymer/expression-scanner.ts (hunk 5 above)
Comment (rictic) [inline src/polymer/expression-scanner.ts]: Also went with `direction` rather than `directionality`
